torch/nn/modules/module.py

```diff
@@ -353,44 +356,83 @@ def arg_error():
             arg_reprs = list(repr(arg) for arg in args)
             for key, val in kwargs.items():
                 arg_reprs.append("{}={}".format(key, val))
-            return ValueError('module.to expects .to(device), .to(dtype) or '
-                              '.to(device, dtype), where dtype is a floating '
-                              'point type, but got .to({})'
+            return ValueError('module.to expects .to(device, non_blocking=False), '
+                              '.to(dtype) or .to(device, dtype=None, non_blocking=False), '
+                              'where dtype is a floating point type, but got .to({})'
                               .format(", ".join(arg_reprs)))
 
         nargs = len(args) + len(kwargs)
-        device = dtype = None
-        if nargs < 1 or nargs > 2:
+        if nargs < 1 or nargs > 3:
             raise arg_error()
+
+        # detect modes by checking first arg
+        if len(args) > 0:
+            arg0 = args[0]
         else:
-            for key, val in kwargs.items():
-                if key == 'dtype':
-                    dtype = kwargs['dtype']
-                elif 'device' in kwargs:
-                    device = kwargs['device']
-                else:
-                    raise arg_error()
-            for arg in args:
-                if isinstance(arg, torch.dtype):
-                    if dtype is not None:
-                        raise arg_error()
-                    dtype = arg
-                else:
-                    if device is not None:
-                        raise arg_error()
-                    device = arg
+            # 'device' has priority because of .to(device, dtype=None, non_blocking=False)
```

Comment:
Aww man, dis argument parsing code ^^

Reply:
yeah I know :/

Reply:
is it possible to move this to C++ so we can just use the same function signatures?  Maybe make an internal function or something that returns a tuple idk.